lib/message_batch.js

```diff
@@ -24,23 +24,65 @@ var Buffer = require('buffer').Buffer;
 
 // The batch message payload should follow:
 // 4 BE bytes number of messages + 4 BE bytes size of message + actual message
+var KAFKA_PADDINGS = {
+  MESSAGE_PADDING_BYTES: 4,
+  BATCH_PADDING_BYTES: 4
+};
+Object.freeze(KAFKA_PADDINGS);
+
 function MessageBatch(size) {
     var self = this;
 
-    self.cachedBuf = new Buffer(size);
-    self.currOffset = 4;
+    self.maxBufferSize = size;
+    self.cachedBuf = new Buffer(self.maxBufferSize);
+    self.currOffset = KAFKA_PADDINGS.BATCH_PADDING_BYTES;
     self.timestamp = new Date().getTime();
     self.numMessages = 0;
-    self.sizeBytes = 4;
+    self.sizeBytes = KAFKA_PADDINGS.BATCH_PADDING_BYTES;
     self.pendingCallbacks = [];
 }
 
+MessageBatch.prototype._throwErrorIfNoCallback = function _throwErrorIfNoCallback(errorMessage, callback) {
+    var err = new Error(errorMessage);
+    if (!callback) {
+        throw err;
+    }
+    callback(err);
+};
+
+MessageBatch.prototype.exceedsBatchSizeBytes = function exceedsBatchSizeBytes(message, callback) {
+    var self = this;
+
+    if (typeof message !== 'string' && !Buffer.isBuffer(message)) {
+        // Not sure what this object is
+        self._throwErrorIfNoCallback('For batching, message must be a string or buffer, not ' + (typeof message), callback);
+        return null;
+    }
+
+    // No single message can be a batch by itself unless it is smaller
+    // than our buffer size
+    return message.length + KAFKA_PADDINGS.BATCH_PADDING_BYTES +
+        KAFKA_PADDINGS.MESSAGE_PADDING_BYTES > self.maxBufferSize;
+};
+
+MessageBatch.prototype.canAddMessageToBatch = function canAddMessageToBatch(message, callback) {
+    var self = this;
+
+    if (typeof message !== 'string' && !Buffer.isBuffer(message)) {
+        // Not sure what this object is
+        self._throwErrorIfNoCallback('For batching, message must be a string or buffer, not ' + (typeof message), callback);
```

Comment:
For format check, I think it's better to do at the beginning of KafkaProducer.prototype.batch and return if fails; If we check here, this message will still be added to batch queue( through it will not succeed since MessageBatch.prototype.addMessage will also check this).